nin/frontend/app/scripts/components/demoplayer.js

```diff
@@ -0,0 +1,108 @@
+const React = require('react');
+const commands = require('../commands');
+const errorExcerpt = require('../utils/errorExcerpt');
+const stackFormat = require('../utils/stackFormat');
+
+class DemoPlayer extends React.Component {
+  constructor(props) {
+    super(props);
+
+    commands.on('resetFlyFlightDynamics', function() {
+      camera.resetFlyFlightDynamics();
+    });
+    commands.on('increaseCameraZoom', function() {
+      camera.deltaFov(-0.5);
+    });
+    commands.on('decreaseCameraZoom', function() {
+      camera.deltaFov(0.5);
+    });
+
+    setTimeout(() => {
+      this.props.demo.resize();
+    });
+
+    this.resizeOverlay = () => {
+      const rect = this.container.getBoundingClientRect();
+      let width = 0;
+      let height = 0;
+      if(rect.width / rect.height < 16 / 9) {
+        width = rect.width;
+        height = width / 16 * 9;
+      } else {
+        height = rect.height;
+        width = height / 9  * 16;
+      }
+      this.overlayElement.style.width = `${width}px`;
+      this.overlayElement.style.height = `${height}px`;
+      this.overlayElement.style.left = `${(rect.width - width) / 2}px`;
+    };
+  }
+
+  toggleFlyAroundMode() {
+    return this.camera.toggleFlyAroundMode();
+  }
+
+  resetFlyFlightDynamics() {
+    return this.camera.resetFlyFlightDynamics();
+  }
+
+  componentDidMount() {
+    this.resizeOverlay();
+
+    window.addEventListener('resize', this.resizeOverlay);
+
+    setTimeout(() => {
+      this.props.demo.start();
+      this.props.demo.music.pause();
+      this.props.demo.jumpToFrame(0);
+    }, 0);
+
+    this.props.demo.setContainer(this.container);
+  }
+
+  componentWillReceiveProps(nextProps) {
+    if (nextProps.fullscreen) {
+      // go to fullscreen
+      document.body.classList.add('fullscreen');
+    } else {
+      // exit fullscreen
+      document.body.classList.remove('fullscreen');
+    }
+    this.props.demo.resize();
+    this.resizeOverlay();
+  }
+
+  render() {
+    return (
+      <div className="floating-viewer-container">
+        { /*cramped formatting because of whitespace: pre */ }
+        { Object.values(this.props.demo.globalJSErrors).map(error => (
+          <div className='global-error-overlay'>
+            <div className='context'>⛄ { error.context } ⛄</div>
+            <div className="bg">
+              <div className="header"><strong>{ error.stack.split('\n')[0] }</strong></div>
+              <div className="excerpt">
+                { errorExcerpt(error, this.props.fileCache) }
+              </div>
+              <div className="stack">
+                { stackFormat(error) }
+              </div>
+            </div>
+          </div>
+        ))}
+        <div className="demo-container" ref={ref => this.container = ref}>
+          <img
+            src="images/overlay.png"
+            className="overlay"
+            style={{
+              display: this.props.showFramingOverlay ? 'block' : 'none'
```

Comment:
I left it as it was because I suspected toggling would be slower with removing the whole img, yes